src/GSoft.Extensions.Http.Authentication.ClientCredentialsGrant/ClientCredentialsTokenEndpointService.cs

```diff
@@ -0,0 +1,96 @@
+﻿using System.Net;
+using IdentityModel.Client;
+using Microsoft.Extensions.Options;
+
+namespace GSoft.Extensions.Http.Authentication.ClientCredentialsGrant;
+
+/// <summary>
+/// Service responsible for retrieving a client credentials grant-based access token from an OAuth 2.0 identity provider.
+/// </summary>
+internal class ClientCredentialsTokenEndpointService : IClientCredentialsTokenEndpointService
+{
+    private readonly IHttpClientFactory _httpClientFactory;
+    private readonly IOptionsMonitor<ClientCredentialsOptions> _optionsMonitor;
+    private readonly IOpenIdConfigurationRetriever _oidcRetriever;
+
+    public ClientCredentialsTokenEndpointService(IHttpClientFactory httpClientFactory, IOptionsMonitor<ClientCredentialsOptions> optionsMonitor, IOpenIdConfigurationRetriever oidcRetriever)
+    {
+        this._httpClientFactory = httpClientFactory;
+        this._optionsMonitor = optionsMonitor;
+        this._oidcRetriever = oidcRetriever;
+    }
+
+    public async Task<ClientCredentialsToken> RequestTokenAsync(string clientName, CancellationToken cancellationToken)
+    {
+        var options = this._optionsMonitor.Get(clientName);
+
+        var metadataEndpoint = await this._oidcRetriever.GetAsync(options.Authority, cancellationToken).ConfigureAwait(false);
+
+        var request = new ClientCredentialsTokenRequest
+        {
+            Address = metadataEndpoint.TokenEndpoint,
+            Scope = options.Scope,
+            ClientId = options.ClientId,
+            ClientSecret = options.ClientSecret,
+            ClientCredentialStyle = ClientCredentialStyle.AuthorizationHeader,
+        };
+
+        var httpClient = this._httpClientFactory.CreateClient(ClientCredentialsConstants.BackchannelHttpClientName);
+
+        // Eventually replace IdentityModel with Microsoft.Identity.Client (MSAL) when their generic authority feature is mature
+        var response = await httpClient.RequestClientCredentialsTokenAsync(request, cancellationToken).ConfigureAwait(false);
+
+        try
+        {
+            ThrowOnIdentityModelError(clientName, response);
+
+            return new ClientCredentialsToken
+            {
+                AccessToken = response.AccessToken,
+                Expiration = response.ExpiresIn == 0 ? DateTimeOffset.MaxValue : DateTimeOffset.UtcNow.AddSeconds(response.ExpiresIn),
```

Comment:
Shouldn't the check be `response.ExpiresIn <= 0` in case the expiration has already passed

Reply:
Makes sense then!